### PR TITLE
primality constraint

### PR DESCRIPTION
**Pull request by robdockins**
Dust off some experimental work on a type-level primality constraint, and push it forward.

This PR implements a primality constraint, which is implemented in the typechecker via a probable prime test.  The type `Z p` is added to the `Field` class under the constraint `prime p`, and `recip` is defined via `EGCD` for concrete values.  For symbolic values, the What4 backend posits the multiplicative inverse via a defining equation.  The SBV backend needs to be enhanced with the same definition equation capability the What4 backend has to complete this work.

### CODE REVIEW

**Comment by robdockins** [comment]
The primality constraint and prime field inverses are now fully implemented.  Remaining tasks are to add some unit tests and documentation.

**Comment by brianhuffman** [comment]
Do the changes to the `Cryptol.Eval.SBV` API in f3e5200d make sense on their own? This might be easier to review if that was split into a separate PR.

**Comment by robdockins** [comment]
> Do the changes to the `Cryptol.Eval.SBV` API in [f3e5200](https://github.com/GaloisInc/cryptol/commit/f3e5200d073009f138fa3264041a9dd3ee05c8da) make sense on their own? This might be easier to review if that was split into a separate PR.

Yeah, that might make sense.  I was planning to also unify some of the What4 and SBV code paths, and maybe all of that makes more sense as separate PR.

File: src/Cryptol/Eval/Concrete/Value.hs [hunk 1]
```diff
@@ -316,6 +317,15 @@ instance Backend Concrete where
   znToInt _ _m x = pure x
   znEq _ _m x y = pure $! x == y
 
+  -- NB: under the precondition that `m` is prime,
+  -- the only values for which no inverse exists are
+  -- congruent to 0 modulo m.
+  znRecip sym m x
+    | r == 0    = raiseError sym DivideByZero
```
Comment (brianhuffman) [inline src/Cryptol/Eval/Concrete/Value.hs:324]: Does `Integer.recipModInteger` return 0 if `x` is 0, or did you mean to test `x == 0` here?

Comment (robdockins) [reply src/Cryptol/Eval/Concrete/Value.hs:324]: `Integer.recipModInteger` returns 0 whenever the input does not have a multiplicative inverse, including for 0.

**Review by brianhuffman — APPROVED** [review]
Looks good to me. At first I was a bit unsure about using a trie to cache primality checks; is the time savings worth the memory cost? But considering the typical use case, it probably is: Users will have a few fixed (probably large) prime constants in their source files, and you'll have to discharge primality constraints on the same numbers every time you use one of the overloaded primitives.

**Pull request merged**